pulse2percept/retina.py

```diff
@@ -1,772 +1,958 @@
+        function are given by self.asymptote, self.slope, and self.shift,
+        respectively.
+
+        Parameters
+        ----------
+        stim : array
+           Temporal signal to process, stim(r,t) in Nanduri et al. (2012).
+
+        Returns
+        -------
+        Rescaled signal, b4(r,t) in Nanduri et al. (2012).
+
+        Notes
+        -----
+        Conversion to TimeSeries is avoided for the sake of speedup.
+        """
+        # use expit (logistic) function for speedup
+        sigmoid = ss.expit((stim.max() - self.shift) / self.slope)
+
+        # avoid division by zero
+        return stim * sigmoid
+
+    def slow_response(self, stim):
+        """Slow response function
+
+        Convolve a stimulus `stim` with a low-pass filter (3-stage gamma)
+        with time constant self.tau_slow.
+        This is Box 5 in Nanduri et al. (2012).
+
+        Parameters
+        ----------
+        stim : array
+           Temporal signal to process, stim(r,t) in Nanduri et al. (2012)
+
+        Returns
+        -------
+        Slow response, b5(r,t) in Nanduri et al. (2012).
+
+        Notes
+        -----
+        This is by far the most computationally involved part of the perceptual
+        sensitivity model.
+        Conversion to TimeSeries is avoided for the sake of speedup.
+        """
+        # No need to zero-pad: fftconvolve already takes care of optimal
+        # kernel/data size
+        conv = signal.fftconvolve(stim, self.gamma_slow, mode='full')
+
+        # Cut off the tail of the convolution to make the output signal match
+        # the dimensions of the input signal.
+        return self.scale_slow * self.tsample * conv[:stim.shape[-1]]
+
+    def calc_layer_current(self, ecs_item, pt_list, layers):
+        """For a given pixel, calculates the effective current for each retinal
+           layer over time
+
+        This function operates at a single-pixel level: It calculates the
+        combined current from all electrodes through a spatial location
+        over time. This calculation is performed per retinal layer.
+
+        Parameters
+        ----------
+        ecs_item: array-like
+            A 2D array specifying the effective current values at a
+            particular spatial location (pixel); one value per retinal
+            layer and electrode.
+            Dimensions: <#layers x #electrodes>
+        pt_list: list
+            A list of PulseTrain `data` containers.
+            Dimensions: <#electrodes x #time points>
+        layers : list
+            List of retinal layers to simulate. Choose from:
+            - 'OFL': optic fiber layer
+            - 'GCL': ganglion cell layer
+            - 'INL': inner nuclear layer
+        """
+        not_supported = np.array([l not in SUPPORTED_LAYERS for l in layers],
+                                 dtype=bool)
+        if any(not_supported):
+            raise ValueError("Acceptable values for `layers` is 'OFL', 'GCL', "
```

Comment:
This is alright, but error handling in the `Electrode` initialization was much more elegant :-)